File: priority_queue/pq.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-


class PQ(object):

    """An implementation of Max Priority Queue"""

    def __init__(self):
        self.pq = [None]
        self.size = 0

    def insert(self, item):
        self.pq.append(item)
        self.size += 1
        self._swim(self.size)

    def del_max(self):
        if not self.size > 0:
            raise IndexError
        rv = self.pq[1]
        self._exch(1, self.size)
        self.pq.pop()
        self.size -= 1
        self._sink(1)
        return rv

    def _exch(self, idx1, idx2):
        self.pq[idx1], self.pq[idx2] = self.pq[idx2], self.pq[idx1]

    def _sink(self, idx):
        while idx*2 < self.size:
            item = self.pq[idx]
            only_one_child = idx*2+1 > self.size
            if only_one_child:
                child = self.pq[idx*2]
                if child <= item:
                    return
                self._exch(idx, idx*2)
                idx = idx*2
            else:
                child1 = self.pq[idx*2]
                child2 = self.pq[idx*2+1]
                if not (child1 > item or child2 > item):
                    return
                bigger_child_idx = idx*2 if child1 > child2 else idx*2+1
                self._exch(idx, bigger_child_idx)
                idx = bigger_child_idx

    def _swim(self, idx):
        while idx > 1:
            item = self.pq[idx]
            parent = self.pq[idx//2]
            if parent >= item:
                return
            self._exch(idx, idx//2)
            idx = idx // 2


if __name__ == "__main__":
    import random
    from pprint import pprint
    pq = PQ()
    numbers = list(range(0, 15))
    random.shuffle(numbers)
    pprint(numbers)
    for n in numbers:
        pq.insert(n)
    pprint(pq.pq)
    while pq.size > 1:
        pprint(pq.del_max())

File: priority_queue/pq.ts
class PQ<T> {
    public pq: (T|null)[];
    constructor () {
        this.pq = [null];
    }

    insert(item: T): void {
        this.pq.push(item);
        this.swim();
    }

    remove_max(): T {
        const rv = this.pq[1];
        this.swap(1, this.pq.length-1);
        this.pq.pop();
        this.sink();
        return <T>rv;
    }

    private swim(): void {
        let idx = this.pq.length-1;
        while(idx > 1) {
            const parent_idx = Math.floor(idx / 2);
            const is_larger_than_parent: boolean = 
                <T>this.pq[idx] > <T>this.pq[parent_idx];
            if(!is_larger_than_parent) return;
            this.swap(idx, parent_idx);
            idx = parent_idx;
        }
    }

    private sink(): void {
        let idx = 1;
        let [child1_idx, child2_idx] = [idx*2, idx*2+1];
        while(child1_idx < this.pq.length - 1) {
            const only_one_child: boolean = child2_idx == this.pq.length - 1;
            if (only_one_child)  {
                if (<T>this.pq[idx] >= <T>this.pq[child1_idx]) 
                    return;
                this.swap(idx, child1_idx);
                idx = child1_idx;
                [child1_idx, child2_idx] = [idx*2, idx*2+1];
            }
            else {
                const child1 = <T>this.pq[child1_idx];
                const child2 = <T>this.pq[child2_idx];
                if (!(child1 > <T>this.pq[idx] || child2 > <T>this.pq[idx]))
                    return;
                const bigger_child_idx: number = child1 > child2 ? child1_idx : child2_idx;
                this.swap(idx, bigger_child_idx);
                idx = bigger_child_idx;
                [child1_idx, child2_idx] = [idx*2, idx*2+1];
            }
        }
    }

    private swap(idx1: number, idx2: number): void {
        const temp = <T>this.pq[idx1];
        this.pq[idx1] = this.pq[idx2];
        this.pq[idx2] = temp;
    }
}

const pq: PQ<number> = new PQ();
for (let i=0; i<10; i++)
    pq.insert(i);
pq.remove_max();
while (pq.pq.length > 1) {
    console.log(pq.remove_max());
}
console.log(pq.pq);